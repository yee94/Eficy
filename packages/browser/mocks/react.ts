var modules = window['React'];

export const Children = modules['Children'];
export const Component = modules['Component'];
export const Fragment = modules['Fragment'];
export const Profiler = modules['Profiler'];
export const PureComponent = modules['PureComponent'];
export const StrictMode = modules['StrictMode'];
export const Suspense = modules['Suspense'];
export const __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED =
  modules['__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED'];
export const cloneElement = modules['cloneElement'];
export const createContext = modules['createContext'];
export const createElement = modules['createElement'];
export const createFactory = modules['createFactory'];
export const createRef = modules['createRef'];
export const forwardRef = modules['forwardRef'];
export const isValidElement = modules['isValidElement'];
export const lazy = modules['lazy'];
export const memo = modules['memo'];
export const startTransition = modules['startTransition'];
export const unstable_act = modules['unstable_act'];
export const useCallback = modules['useCallback'];
export const useContext = modules['useContext'];
export const useDebugValue = modules['useDebugValue'];
export const useDeferredValue = modules['useDeferredValue'];
export const useEffect = modules['useEffect'];
export const useId = modules['useId'];
export const useImperativeHandle = modules['useImperativeHandle'];
export const useInsertionEffect = modules['useInsertionEffect'];
export const useLayoutEffect = modules['useLayoutEffect'];
export const useMemo = modules['useMemo'];
export const useReducer = modules['useReducer'];
export const useRef = modules['useRef'];
export const useState = modules['useState'];
export const useSyncExternalStore = modules['useSyncExternalStore'];
export const useTransition = modules['useTransition'];
export const version = modules['version'];

export default modules;
