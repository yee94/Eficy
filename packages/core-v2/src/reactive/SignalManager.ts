import { createSignal, createComputed } from 'reactjs-signal';
import { effect } from 'alien-signals';
import { injectable } from 'tsyringe';
import type { ISignalManager, ISignal } from '../interfaces';

/**
 * 信号实现类
 */
class SignalImpl<T> implements ISignal<T> {
  constructor(private signalRef: ReturnType<typeof createSignal<T>>) {}

  get(): T {
    return this.signalRef();
  }

  set(value: T): void {
    this.signalRef(value);
  }

  update(fn: (prev: T) => T): void {
    this.signalRef(fn);
  }

  subscribe(listener: (value: T) => void): () => void {
    // reactjs-signal 基于 alien-signals，可以通过 effect 来监听变化
    return effect(() => {
      listener(this.signalRef());
    });
  }
}

/**
 * 计算信号实现类
 */
class ComputedSignalImpl<T> implements ISignal<T> {
  constructor(private computedRef: ReturnType<typeof createComputed<T>>) {}

  get(): T {
    return this.computedRef();
  }

  set(): void {
    throw new Error('Cannot set value of computed signal');
  }

  update(): void {
    throw new Error('Cannot update value of computed signal');
  }

  subscribe(listener: (value: T) => void): () => void {
    return effect(() => {
      listener(this.computedRef());
    });
  }
}

/**
 * 响应式信号管理器
 * 基于reactjs-signal实现，替代mobx
 */
@injectable()
export class SignalManager implements ISignalManager {
  /**
   * 创建响应式信号
   */
  createSignal<T>(initialValue: T): ISignal<T> {
    const signalRef = createSignal(initialValue);
    return new SignalImpl(signalRef);
  }

  /**
   * 创建计算信号
   */
  createComputed<T>(fn: () => T): ISignal<T> {
    const computedRef = createComputed(fn);
    return new ComputedSignalImpl(computedRef);
  }

  /**
   * 创建副作用
   */
  createEffect(fn: () => void | (() => void)): () => void {
    return effect(fn);
  }

  /**
   * 批量更新信号
   */
  batch(fn: () => void): void {
    // reactjs-signal 基于 alien-signals，会自动批量更新
    fn();
  }
} 