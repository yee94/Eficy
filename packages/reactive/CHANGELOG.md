# @eficy/reactive

## 1.0.1

### Patch Changes

- 155c510: 这是一个示例 changeset，展示了如何描述更改。

  - 修复了响应式系统的类型错误
  - 优化了 React 集成的性能
  - 更新了 JSX 运行时的实现

- b74eab7: feature: @eficy/core-jsx is now available! 🎉
