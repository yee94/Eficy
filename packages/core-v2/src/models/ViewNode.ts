import { IView } from '../interface';
import { Vmo } from '@vmojs/base';
import { Field } from '@vmojs/base/bundle';
import {
  cloneDeep,
  deleteObjectField,
  forEachDeep,
  generateUid,
  isArray,
  isEficyView,
  isObject,
  mapDeep,
  MERGE_WAY,
  mergeWith,
} from '../utils';
import { action, computed, observable } from 'mobx';
import Config from '../constants/Config';
import UnEnumerable from '../utils/decorators/UnEnumerable';
import { CSSProperties } from 'react';

type IComponentsModels = Record<string, new (...args: any) => ViewNode>;
export type ExtendsViewNode = ViewNode & any;

export default class ViewNode extends Vmo implements IView {
  public static readonly solidField = ['#', '#view', '#restProps'];
  public static isInstance(someObject: any): boolean {
    return someObject.__proto__.constructor.name === ViewNode.name;
  }

  @UnEnumerable
  private readonly componentModels: IComponentsModels;

  constructor(data: IView, componentModels: IComponentsModels = {}) {
    super();

    this.componentModels = componentModels || {};
    const foundModel = this.componentModels[data['#view']];

    if (foundModel && ViewNode.isInstance(this)) {
      // @ts-ignore
      return new foundModel(data, componentModels);
    } else {
      this.load(data);
    }
  }

  @Field
  public '#view': string;
  @Field
  public '#': string;

  @Field
  @observable.ref
  public '#children': ViewNode[];

  @Field
  public '#staticProps': Record<string, any> = {};

  @Field
  @observable
  public '#if': boolean;

  @Field
  @observable
  public '#content': string;

  @observable
  public '#restProps': Record<string, any>;

  @Field
  @observable
  public className: string;

  @Field
  @observable
  public style: CSSProperties;

  @action
  protected load(data: IView): this {
    if (!data) {
      return this;
    }
    if (!data['#']) {
      data['#'] = generateUid();
    }

    if (!('#if' in data)) {
      data['#if'] = true;
    }

    this.update(data, true);

    return this;
  }

  @computed
  public get viewDataMap(): Record<string, ExtendsViewNode> {
    const viewMaps = { [this['#']]: this };
    const extendViewMap = (viewNode: ViewNode) => {
      const childMap = viewNode.viewDataMap;
      if (childMap) {
        Object.assign(viewMaps, childMap);
      }
    };
    (this['#children'] || []).forEach(extendViewMap);

    forEachDeep(this['#restProps'], (optionValue) => {
      if (optionValue instanceof ViewNode) {
        extendViewMap(optionValue);
      }

      return optionValue;
    });

    return viewMaps;
  }

  @action
  private transformViewNode(data: IView) {
    return mapDeep(
      data,
      (value: IView, path) => {
        if (!path) {
          return value;
        }
        if (isEficyView(value) && !(value instanceof ViewNode)) {
          return new ViewNode(value, this.componentModels);
        }
        return value;
      },
      {
        exceptFns: [...Config.loopExceptFns, (obj) => obj['#view'] === 'Eficy'],
      },
    );
  }

  @action
  public overwrite(data: IView) {
    [...Object.keys(this), ...Object.keys(this['#restProps'])].forEach((key) => {
      if (ViewNode.solidField.includes(key)) {
        return;
      }
      const originValue = ViewNode.prototype[key];
      switch (typeof originValue) {
        case 'function':
        case 'object':
          this[key] = cloneDeep(originValue);
          break;
        case 'undefined':
          deleteObjectField(this, key);
          break;
        default:
          this[key] = originValue;
      }
    });
    // @ts-ignore
    this['#restProps'] = undefined;
    deleteObjectField(this, '#children');

    this.load(data);

    return this;
  }

  @action
  public update(data: IView, isInit = false): this {
    if (!this['#restProps']) {
      this['#restProps'] = {};
      Object.defineProperty(this, '#restProps', {
        enumerable: false,
      });
    }

    data = this.transformViewNode(data);

    Object.keys(data).forEach((key) => {
      if (!isInit && ViewNode.solidField.includes(key)) {
        return;
      }

      // @ts-ignore
      if (this.__proto__.hasOwnProperty(key)) {
        this[key] = data[key];
      } else {
        if (!this.hasOwnProperty(key)) {
          this['#restProps'][key] = data[key];
          Object.defineProperty(this, key, {
            enumerable: true,
            configurable: true,
            get() {
              return this['#restProps'][key];
            },
            set(val) {
              this['#restProps'][key] = val;
            },
          });
        } else {
          if (isObject(data[key])) {
            this[key] = mergeWith(this[key], data[key], MERGE_WAY.REPLACE);
          } else {
            this[key] = data[key];
          }
        }
      }
    });

    return this;
  }

  public forEachChild(cb: (child: ViewNode) => void) {
    if (isArray(this['#children'])) {
      this['#children'].forEach((child) => {
        cb(child);
        if (isArray(child['#children'])) {
          child.forEachChild(cb);
        }
      });
    }
  }
}
