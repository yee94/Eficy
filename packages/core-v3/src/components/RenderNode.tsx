import { useObserver } from '@eficy/reactive-react';
import type { ComponentType, FC } from 'react';
import { createElement, memo } from 'react';
import { ErrorBoundary } from 'react-error-boundary';
import type { IRenderNodeProps } from '../interfaces';

// 自定义错误回退组件
const ErrorFallback = ({ error, resetErrorBoundary }: { error: Error; resetErrorBoundary: () => void }) => {
  return (
    <div style={{ color: 'red', border: '1px solid red', padding: '8px', margin: '4px' }}>
      <h4>Something went wrong</h4>
      <details>
        <summary>Error details</summary>
        <pre>{error.message}</pre>
        <pre>{error.stack}</pre>
      </details>
      <button onClick={resetErrorBoundary} style={{ marginTop: '8px' }}>
        Try again
      </button>
    </div>
  );
};

// 主渲染组件 - 使用useObserver(view)模式来追踪响应式数据
const RenderNodeInner: FC<IRenderNodeProps> = ({ eficyNode, componentMap = {}, childrenMap }: IRenderNodeProps) => {
  if (!childrenMap || !eficyNode) {
    throw new Error('childrenMap and eficyNode are required');
  }
  // 正确地在组件顶层调用useObserver hook
  const renderResult = useObserver(() => {
    const componentName = eficyNode['#view'];
    const shouldRender = eficyNode.shouldRender;
    const props = eficyNode.props;

    // 检查是否应该渲染
    if (!shouldRender) {
      return null;
    }

    const Component = componentMap[componentName] as ComponentType<any>;

    // 组件不存在的错误处理
    if (!Component) {
      console.error(`❌ Component "${componentName}" not found in componentMap`);
      return (
        <div style={{ color: 'red', background: '#ffe6e6', padding: '8px', border: '1px solid red' }}>
          Component "{componentName}" not found
        </div>
      );
    }
    const children = (() => {
      if (Array.isArray(eficyNode.children)) {
        return eficyNode.children.map((child) => childrenMap.get(child.id));
      }
      return eficyNode.children;
    })();

    // 创建最终props
    const finalProps = {
      ...props,
      children,
    };

    // 如果Component是字符串（原生HTML标签）
    if (typeof Component === 'string') {
      return createElement(Component, finalProps);
    }

    // 如果Component是React组件
    return <Component {...finalProps} />;
  });

  // 返回useObserver的结果
  return renderResult;
};

// 使用memo优化的RenderNode
const RenderNode = memo(
  (props: IRenderNodeProps) => {
    // 使用react-error-boundary替代自定义ErrorBoundary
    return (
      <ErrorBoundary
        FallbackComponent={ErrorFallback}
        onError={(error, info) => {
          console.error('💥 RenderNode Error:', error, info);
        }}
      >
        <RenderNodeInner {...props} />
      </ErrorBoundary>
    );
  },
  (prevProps, nextProps) => {
    // 自定义比较函数，只有viewNode发生变化时才重新渲染
    return prevProps.eficyNode === nextProps.eficyNode && prevProps.componentMap === nextProps.componentMap;
  },
);

RenderNode.displayName = 'RenderNode';

export default RenderNode;
