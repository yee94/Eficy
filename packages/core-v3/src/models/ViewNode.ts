import { nanoid } from 'nanoid';
import { observable, computed, action, ObservableClass, makeObservable } from '@eficy/reactive';
import { isFunction } from 'lodash';
import { setOmit } from '../utils/common';
import type { IViewData } from '../interfaces';
import type { ReactElement } from 'react';

// 框架特殊字段，不会传递给组件
const FRAMEWORK_FIELDS = new Set(['#', '#view', '#children', '#content', '#if', '#staticProps']);

export default class ViewNode {
  // 唯一标识
  public readonly id: string = nanoid();

  // 核心字段
  @observable
  public '#' = '';

  @observable
  public '#view' = 'div';

  @observable
  public '#children': ViewNode[] = [];

  @observable
  public '#content'?: string | ReactElement;

  @observable
  public '#if'?: boolean | (() => boolean) = true;

  // 动态属性存储
  @observable
  private dynamicProps: Record<string, any> = {};
  
  // 子模型映射
  @observable
  public models: Record<string, ViewNode> = {};

  constructor(data: IViewData) {
    this.load(data);
    makeObservable(this);
  }

  /**
   * 加载ViewData数据
   */
  @action
  private load(data: IViewData): void {
    console.log('📥 ViewNode.load called:', { 
      currentId: this['#'], 
      newData: data 
    });
    // 设置核心字段
    this['#'] = data['#'] || this.id;
    this['#view'] = data['#view'] || 'div';
    this['#content'] = data['#content'];
    this['#if'] = data['#if'] !== undefined ? data['#if'] : true;

    // 处理子节点
    if (data['#children']) {
      const children = data['#children'].map((childData) => new ViewNode(childData));
      this['#children'] = children;
      
      // 更新模型映射
      this.models = {};
      children.forEach(child => {
        if (child['#']) {
          this.models[child['#']] = child;
        }
      });
    }

    // 设置其他属性
    const otherProps = setOmit(data, FRAMEWORK_FIELDS);
    this.dynamicProps = { ...otherProps };
    console.log('✅ ViewNode.load completed:', { 
      id: this['#'], 
      view: this['#view'],
      content: this['#content'],
      dynamicProps: this.dynamicProps
    });
  }

  /**
   * 计算最终传递给组件的props
   */
  @computed
  get props(): Record<string, any> {
    const props: Record<string, any> = { ...this.dynamicProps };

    // 处理 #content -> children
    if (this['#content'] !== undefined) {
      props.children = this['#content'];
    }

    // 如果有子节点，children 应该是子节点的渲染结果
    if (this['#children'] && this['#children'].length > 0) {
      // 这里只返回子节点数组，实际渲染由 RenderNode 处理
      props.children = this['#children'];
    }

    return props;
  }

  /**
   * 判断是否应该渲染
   */
  @computed
  get shouldRender(): boolean {
    const condition = this['#if'];

    if (condition === undefined || condition === null) {
      return true;
    }

    if (isFunction(condition)) {
      return (condition as () => boolean)();
    }

    return Boolean(condition);
  }

  /**
   * 更新字段值
   */
  @action
  updateField(key: string, value: any): void {
          console.log('🔄 ViewNode.updateField called:', { 
        id: this['#'], 
        field: key, 
        oldValue: FRAMEWORK_FIELDS.has(key) ? (this as any)[key] : this.dynamicProps[key],
        newValue: value 
      });
    if (FRAMEWORK_FIELDS.has(key)) {
      // 更新框架字段
      (this as any)[key] = value;
      console.log(`✅ Updated system field ${key} to:`, value);
    } else {
      // 更新动态属性 - 使用不可变更新
      this.dynamicProps = {
        ...this.dynamicProps,
        [key]: value,
      };
      console.log(`✅ Updated dynamic field ${key} to:`, value);
    }
  }

  /**
   * 添加子节点
   */
  @action
  addChild(child: ViewNode): void {
    this['#children'] = [...this['#children'], child];
    
    // 更新模型映射
    if (child['#']) {
      this.models = { ...this.models, [child['#']]: child };
    }
  }

  /**
   * 移除子节点
   */
  @action
  removeChild(childId: string): void {
    this['#children'] = this['#children'].filter((child) => child['#'] !== childId);
    
    // 更新模型映射
    if (childId && this.models[childId]) {
      const newModels = { ...this.models };
      delete newModels[childId];
      this.models = newModels;
    }
  }

  /**
   * 查找子节点
   */
  findChild(childId: string): ViewNode | null {
    return this['#children'].find((child) => child['#'] === childId) || null;
  }

  /**
   * 获取视图数据映射
   */
  @computed
  get viewDataMap(): Record<string, ViewNode> {
    const result: Record<string, ViewNode> = { [this['#']]: this };
    
    // 递归合并子节点的 viewDataMap
    this['#children'].forEach(child => {
      Object.assign(result, child.viewDataMap);
    });
    
    return result;
  }
  
  /**
   * 序列化为JSON
   */
  toJSON(): IViewData {
    const result: IViewData = {
      '#': this['#'],
      '#view': this['#view'],
      ...this.dynamicProps,
    };

    if (this['#content'] !== undefined) {
      result['#content'] = this['#content'];
    }

    if (this['#if'] !== true) {
      result['#if'] = this['#if'];
    }

    if (this['#children'].length > 0) {
      result['#children'] = this['#children'].map((child) => child.toJSON());
    }

    return result;
  }

  /**
   * 从JSON创建ViewNode
   */
  static fromJSON(data: IViewData): ViewNode {
    return new ViewNode(data);
  }

  /**
   * 更新节点数据
   */
  @action
  update(data: IViewData): void {
    console.log('📥 ViewNode.update called:', { 
      currentId: this['#'], 
      newData: data 
    });
    // 更新核心字段
    if (data['#view'] !== undefined) this['#view'] = data['#view'];
    if (data['#content'] !== undefined) this['#content'] = data['#content'];
    if (data['#if'] !== undefined) this['#if'] = data['#if'];

    // 更新动态属性
    const otherProps = setOmit(data, FRAMEWORK_FIELDS);
    this.dynamicProps = { ...this.dynamicProps, ...otherProps };

    // 更新子节点
    if (data['#children']) {
      // 查找并更新现有子节点，添加新子节点
      data['#children'].forEach((childData) => {
        const childId = childData['#'];
        if (childId) {
          const existingChild = this.findChild(childId);
          if (existingChild) {
            existingChild.update(childData);
          } else {
            this.addChild(new ViewNode(childData));
          }
        } else {
          this.addChild(new ViewNode(childData));
        }
      });
      
      // 移除不再存在的子节点
      const updatedChildIds = new Set(data['#children']
        .map(childData => childData['#'])
        .filter(Boolean));
        
      this['#children']
        .filter(child => child['#'] && !updatedChildIds.has(child['#']))
        .forEach(child => this.removeChild(child['#']));
    }
    console.log('✅ ViewNode.update completed:', { 
      id: this['#'], 
      view: this['#view'],
      content: this['#content'],
      dynamicProps: this.dynamicProps
    });
  }

  /**
   * 完全覆盖节点数据
   */
  @action
  overwrite(data: IViewData): void {
    console.log('📥 ViewNode.overwrite called:', { 
      currentId: this['#'], 
      newData: data 
    });
    // 重置所有数据
    this.load(data);
    console.log('✅ ViewNode.overwrite completed:', { 
      id: this['#'], 
      view: this['#view'],
      content: this['#content'],
      dynamicProps: this.dynamicProps
    });
  }
}
