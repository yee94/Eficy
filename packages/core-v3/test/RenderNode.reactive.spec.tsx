import React from 'react'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import { signal } from '@eficy/reactive'
import RenderNode from '../src/components/RenderNode'
import ViewNode from '../src/models/ViewNode'
import type { IViewData } from '../src/interfaces'

describe('RenderNode - Reactive Capabilities', () => {
  const mockComponentMap = {
    div: 'div',
    span: 'span',
    button: 'button',
    h1: 'h1',
    CustomComponent: ({ children, ...props }: any) => (
      <div data-testid="custom-component" {...props}>
        {children}
      </div>
    )
  }

  describe('Reactive Property Updates', () => {
    it('should reactively update when ViewNode properties change', async () => {
      const viewData: IViewData = {
        '#': 'test-node',
        '#view': 'div',
        className: 'initial-class',
        '#content': 'Initial Content'
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={mockComponentMap} />)
      
      // 验证初始渲染
      expect(screen.getByText('Initial Content')).toBeInTheDocument()
      expect(screen.getByText('Initial Content')).toHaveClass('initial-class')
      
      // 响应式更新属性
      viewNode.updateField('className', 'updated-class')
      viewNode.updateField('#content', 'Updated Content')
      
      // 验证响应式更新
      await waitFor(() => {
        expect(screen.getByText('Updated Content')).toBeInTheDocument()
        expect(screen.getByText('Updated Content')).toHaveClass('updated-class')
      })
    })

    it('should reactively update component type', async () => {
      const viewData: IViewData = {
        '#': 'dynamic-component',
        '#view': 'div',
        '#content': 'Content'
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={mockComponentMap} />)
      
      // 验证初始为 div
      expect(screen.getByText('Content').tagName).toBe('DIV')
      
      // 更新组件类型
      viewNode.updateField('#view', 'span')
      
      // 验证响应式更新为 span
      await waitFor(() => {
        expect(screen.getByText('Content').tagName).toBe('SPAN')
      })
    })

    it('should reactively update custom component props', async () => {
      const viewData: IViewData = {
        '#': 'custom-node',
        '#view': 'CustomComponent',
        'data-value': 'initial',
        '#content': 'Custom Content'
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={mockComponentMap} />)
      
      // 验证初始渲染
      const customComponent = screen.getByTestId('custom-component')
      expect(customComponent).toHaveAttribute('data-value', 'initial')
      expect(customComponent).toHaveTextContent('Custom Content')
      
      // 响应式更新属性
      viewNode.updateField('data-value', 'updated')
      viewNode.updateField('#content', 'Updated Custom Content')
      
      // 验证响应式更新
      await waitFor(() => {
        expect(customComponent).toHaveAttribute('data-value', 'updated')
        expect(customComponent).toHaveTextContent('Updated Custom Content')
      })
    })
  })

  describe('Reactive Children Updates', () => {
    it('should reactively update when children are added', async () => {
      const parentData: IViewData = {
        '#': 'parent',
        '#view': 'div',
        '#children': [
          {
            '#': 'child1',
            '#view': 'span',
            '#content': 'Child 1'
          }
        ]
      }

      const parentNode = new ViewNode(parentData)
      
      render(<RenderNode viewNode={parentNode} componentMap={mockComponentMap} />)
      
      // 验证初始子节点
      expect(screen.getByText('Child 1')).toBeInTheDocument()
      expect(screen.queryByText('Child 2')).not.toBeInTheDocument()
      
      // 添加新子节点
      const newChild = new ViewNode({
        '#': 'child2',
        '#view': 'span',
        '#content': 'Child 2'
      })
      
      parentNode.addChild(newChild)
      
      // 验证响应式添加
      await waitFor(() => {
        expect(screen.getByText('Child 2')).toBeInTheDocument()
      })
    })

    it('should reactively update when children are removed', async () => {
      const parentData: IViewData = {
        '#': 'parent',
        '#view': 'div',
        '#children': [
          {
            '#': 'child1',
            '#view': 'span',
            '#content': 'Child 1'
          },
          {
            '#': 'child2',
            '#view': 'span',
            '#content': 'Child 2'
          }
        ]
      }

      const parentNode = new ViewNode(parentData)
      
      render(<RenderNode viewNode={parentNode} componentMap={mockComponentMap} />)
      
      // 验证初始子节点
      expect(screen.getByText('Child 1')).toBeInTheDocument()
      expect(screen.getByText('Child 2')).toBeInTheDocument()
      
      // 移除子节点
      parentNode.removeChild('child1')
      
      // 验证响应式移除
      await waitFor(() => {
        expect(screen.queryByText('Child 1')).not.toBeInTheDocument()
        expect(screen.getByText('Child 2')).toBeInTheDocument()
      })
    })

    it('should reactively update nested children', async () => {
      const rootData: IViewData = {
        '#': 'root',
        '#view': 'div',
        '#children': [
          {
            '#': 'level1',
            '#view': 'div',
            '#children': [
              {
                '#': 'level2',
                '#view': 'span',
                '#content': 'Deep Content'
              }
            ]
          }
        ]
      }

      const rootNode = new ViewNode(rootData)
      
      render(<RenderNode viewNode={rootNode} componentMap={mockComponentMap} />)
      
      // 验证深层内容
      expect(screen.getByText('Deep Content')).toBeInTheDocument()
      
      // 更新深层子节点
      const level1Node = rootNode.findChild('level1')
      const level2Node = level1Node?.findChild('level2')
      level2Node?.updateField('#content', 'Updated Deep Content')
      
      // 验证响应式更新
      await waitFor(() => {
        expect(screen.getByText('Updated Deep Content')).toBeInTheDocument()
        expect(screen.queryByText('Deep Content')).not.toBeInTheDocument()
      })
    })
  })

  describe('Conditional Rendering Reactivity', () => {
    it('should reactively show/hide based on #if condition', async () => {
      const viewData: IViewData = {
        '#': 'conditional',
        '#view': 'div',
        '#content': 'Conditional Content',
        '#if': true
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={mockComponentMap} />)
      
      // 验证初始显示
      expect(screen.getByText('Conditional Content')).toBeInTheDocument()
      
      // 隐藏元素
      viewNode.updateField('#if', false)
      
      // 验证响应式隐藏
      await waitFor(() => {
        expect(screen.queryByText('Conditional Content')).not.toBeInTheDocument()
      })
      
      // 重新显示元素
      viewNode.updateField('#if', true)
      
      // 验证响应式显示
      await waitFor(() => {
        expect(screen.getByText('Conditional Content')).toBeInTheDocument()
      })
    })

    it('should reactively update with function-based #if condition', async () => {
      const showCondition = signal(true)
      
      const viewData: IViewData = {
        '#': 'functional-conditional',
        '#view': 'div',
        '#content': 'Function Conditional',
        '#if': () => showCondition()
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={mockComponentMap} />)
      
      // 验证初始显示
      expect(screen.getByText('Function Conditional')).toBeInTheDocument()
      
      // 通过 signal 改变条件
      showCondition(false)
      
      // 验证响应式隐藏
      await waitFor(() => {
        expect(screen.queryByText('Function Conditional')).not.toBeInTheDocument()
      })
      
      // 重新显示
      showCondition(true)
      
      // 验证响应式显示
      await waitFor(() => {
        expect(screen.getByText('Function Conditional')).toBeInTheDocument()
      })
    })
  })

  describe('Performance and Optimization', () => {
    it('should only re-render affected components', async () => {
      const renderSpy = vi.fn()
      
      const SpyComponent = ({ children, ...props }: any) => {
        renderSpy()
        return <div data-testid="spy-component" {...props}>{children}</div>
      }
      
      const componentMapWithSpy = {
        ...mockComponentMap,
        SpyComponent
      }
      
      const parentData: IViewData = {
        '#': 'parent',
        '#view': 'div',
        '#children': [
          {
            '#': 'child1',
            '#view': 'SpyComponent',
            '#content': 'Child 1'
          },
          {
            '#': 'child2',
            '#view': 'span',
            '#content': 'Child 2'
          }
        ]
      }

      const parentNode = new ViewNode(parentData)
      
      render(<RenderNode viewNode={parentNode} componentMap={componentMapWithSpy} />)
      
      const initialRenderCount = renderSpy.mock.calls.length
      
      // 只更新 child2，不应该重新渲染 child1 的 SpyComponent
      const child2 = parentNode.findChild('child2')
      child2?.updateField('#content', 'Updated Child 2')
      
      await waitFor(() => {
        expect(screen.getByText('Updated Child 2')).toBeInTheDocument()
      })
      
      // SpyComponent 不应该重新渲染
      expect(renderSpy.mock.calls.length).toBe(initialRenderCount)
    })

    it('should handle rapid updates efficiently', async () => {
      const viewData: IViewData = {
        '#': 'rapid-update',
        '#view': 'div',
        '#content': 'Initial'
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={mockComponentMap} />)
      
      // 快速连续更新
      const updates = ['Update 1', 'Update 2', 'Update 3', 'Final Update']
      
      const startTime = performance.now()
      
      updates.forEach((update, index) => {
        setTimeout(() => {
          viewNode.updateField('#content', update)
        }, index * 10)
      })
      
      // 验证最终状态
      await waitFor(() => {
        expect(screen.getByText('Final Update')).toBeInTheDocument()
      }, { timeout: 1000 })
      
      const endTime = performance.now()
      expect(endTime - startTime).toBeLessThan(500) // 应该在合理时间内完成
    })
  })

  describe('Error Handling in Reactive Context', () => {
    it('should handle errors in reactive updates gracefully', async () => {
      const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {})
      
      const ErrorComponent = () => {
        throw new Error('Component Error')
      }
      
      const errorComponentMap = {
        ...mockComponentMap,
        ErrorComponent
      }
      
      const viewData: IViewData = {
        '#': 'error-test',
        '#view': 'div',
        '#content': 'Safe Content'
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={errorComponentMap} />)
      
      // 验证正常渲染
      expect(screen.getByText('Safe Content')).toBeInTheDocument()
      
      // 更新为错误组件
      viewNode.updateField('#view', 'ErrorComponent')
      
      // 验证错误边界处理
      await waitFor(() => {
        expect(screen.getByText('Something went wrong')).toBeInTheDocument()
      })
      
      consoleSpy.mockRestore()
    })

    it('should recover from errors when component is fixed', async () => {
      const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {})
      
      const ErrorComponent = () => {
        throw new Error('Component Error')
      }
      
      const errorComponentMap = {
        ...mockComponentMap,
        ErrorComponent
      }
      
      const viewData: IViewData = {
        '#': 'recovery-test',
        '#view': 'ErrorComponent'
      }

      const viewNode = new ViewNode(viewData)
      
      render(<RenderNode viewNode={viewNode} componentMap={errorComponentMap} />)
      
      // 验证错误状态
      expect(screen.getByText('Something went wrong')).toBeInTheDocument()
      
      // 修复组件
      viewNode.updateField('#view', 'div')
      viewNode.updateField('#content', 'Recovered Content')
      
      // 点击 Try again 按钮
      const tryAgainButton = screen.getByText('Try again')
      tryAgainButton.click()
      
      // 验证恢复
      await waitFor(() => {
        expect(screen.getByText('Recovered Content')).toBeInTheDocument()
        expect(screen.queryByText('Something went wrong')).not.toBeInTheDocument()
      })
      
      consoleSpy.mockRestore()
    })
  })
})