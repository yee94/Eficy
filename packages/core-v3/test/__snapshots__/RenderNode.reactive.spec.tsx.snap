// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RenderNode - Reactive Capabilities > Conditional Rendering Reactivity > should reactively show/hide based on #if condition 1`] = `
{
  "totalRenderNodes": 1,
}
`;

exports[`RenderNode - Reactive Capabilities > Conditional Rendering Reactivity > should reactively update with function-based #if condition 1`] = `
{
  "totalRenderNodes": 1,
}
`;

exports[`RenderNode - Reactive Capabilities > Error Handling in Reactive Context > should handle errors in reactive updates gracefully 1`] = `
{
  "totalRenderNodes": 1,
}
`;

exports[`RenderNode - Reactive Capabilities > Error Handling in Reactive Context > should recover from errors when component is fixed 1`] = `
{
  "totalRenderNodes": 1,
}
`;

exports[`RenderNode - Reactive Capabilities > Performance and Optimization > should handle rapid updates efficiently 1`] = `
{
  "totalRenderNodes": 1,
}
`;

exports[`RenderNode - Reactive Capabilities > Performance and Optimization > should only re-render affected components 1`] = `
{
  "totalRenderNodes": 3,
}
`;

exports[`RenderNode - Reactive Capabilities > Reactive Children Updates > should reactively update nested children 1`] = `
{
  "totalRenderNodes": 3,
}
`;

exports[`RenderNode - Reactive Capabilities > Reactive Children Updates > should reactively update when children are added 1`] = `
{
  "totalRenderNodes": 2,
}
`;

exports[`RenderNode - Reactive Capabilities > Reactive Children Updates > should reactively update when children are removed 1`] = `
{
  "totalRenderNodes": 3,
}
`;

exports[`RenderNode - Reactive Capabilities > Reactive Property Updates > should reactively update component type 1`] = `
{
  "totalRenderNodes": 1,
}
`;

exports[`RenderNode - Reactive Capabilities > Reactive Property Updates > should reactively update custom component props 1`] = `
{
  "totalRenderNodes": 1,
}
`;

exports[`RenderNode - Reactive Capabilities > Reactive Property Updates > should reactively update when EficyNode properties change 1`] = `
{
  "totalRenderNodes": 1,
}
`;
